### PR TITLE
GC: Add a "GC Feature Matrix" persisted object for controlling feature exposure over a document's lifetime

### PR DESCRIPTION
**Pull request by markfields**
## Description

Alternative approach to #13985 after we decided it was too spooky to take a runtime dependency on our version number format (plus potential for bugs in version comparison code).

This introduces a concept of "GC Feature Matrix" to the GC Metadata that we can use to indicate per file (immutable for its lifetime) the level of support the file has for different features.  This will allow us to retroactively disable a feature for old files in case we need to (e.g. if Loop finds a GC-blocking bug during Tombstone phase they can't/won't fix in a self-repairing way, or any such bug during Sweep phase).

For now add a specific dimension there `tombstoneGeneration` which is supplied by the app via an undocumented ContainerRuntimeOption, for the purpose described above.  We will add one for sweep (replacing `SweepEnabled`) next, and may also migrate GC Enabled determination to this instead of using GC Feature.

Fixes [AB#3323](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/3323)

### CODE REVIEW

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts [hunk 1]
```diff
@@ -602,6 +625,52 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
...
+					category: "generic",
+					gcEnforcementDisabled: true,
+				},
+				{
+					eventName:
+						"fluid:telemetry:ContainerRuntime:GarbageCollector:SweepReadyObject_Loaded",
+				},
+			],
+			async function () {
+				// Note: The Summarizers in this test don't use the "future" GC option - it only matters for the interactive client
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:644]: Hmmm... I'd like to test that I can persist the value and then use a different one later, rather than depending on undefined.  Which makes me realize - I'm not persisting the given value!  Will fix tomorrow.

Also, are Summarizer clients given the same RuntimeOptions as the main clients?  I assume so - I mean, it must be, we've been using options for GC for a while.

File: packages/runtime/container-runtime/src/garbageCollectionHelpers.ts [hunk 2]
```diff
@@ -35,3 +38,19 @@ export function sendGCUnexpectedUsageEvent(
...
+ * with a different value will not be subjected to GC enforcement.
+ * @param persistedValue - The persisted feature support value
+ * @param currentValue - The current app-provided feature support value
+ * @returns true if GC Enforcement (Fail on Tombstone load/usage, GC Sweep) should be disabled
+ */
+export function shouldDisableGcEnforcement(
+	persistedValue: number | undefined,
+	currentValue: number | undefined,
+): boolean {
+	return persistedValue !== currentValue;
```
Comment (markfields) [inline packages/runtime/container-runtime/src/garbageCollectionHelpers.ts:55]: This is fine for now, but later I think we should change `undefined` persisted value to be mapped to `0` first, then compare with the incoming current value from the option. See https://dev.azure.com/fluidframework/internal/_workitems/edit/3424 for more details.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 3]
```diff
@@ -1032,6 +1034,11 @@ export class ContainerRuntime
 	 */
 	private nextSummaryNumber: number;
 
+	/**
+	 * If true, loading or using a Tombstoned object should merely log, not fail
+	 */
+	public readonly disableGcTombstoneEnforcement: boolean;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1040]: Can we move this to garbage collector and expose it on IGarabgeCollector. That way this logic remains encapsulated in garbage collector.
This can then become a getter than gets the value from garbage collector.

File: packages/runtime/container-runtime/src/garbageCollectionHelpers.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/garbageCollectionHelpers.ts:55]: Not sure I follow what's described in 3424 but why not return false if persistedValue is undefined? Before applications start passing the GC option, this would be undefined and tombstone will be enabled right?

File: packages/runtime/container-runtime/src/summaryFormat.ts [hunk 4]
```diff
@@ -97,15 +97,35 @@ export interface ICreateContainerMetadata {
...
+
+	/**
+	 * Numerical indication of feature support as of file creation time, used to determine feature availability over time.
+	 * This info may come from multiple sources (FF code, config service, app via Container Runtime Options),
+	 * and pertains to aspects of the document that may be fixed for its lifetime.
+	 *
+	 * For each dimension, if the persisted value is less than the currently provided value,
+	 * then this file does not support the corresponding feature as currently implemented.
+	 */
+	readonly gcFeatureSupportInfo?: GcFeatureSupportDimensions;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summaryFormat.ts:128]:
```suggestion
	readonly gcFeaturesSupportInfo?: GcFeatureSupportDimensions;
```

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/summaryFormat.ts:128]: `features` indicating multiple feature support.

Comment (agarwal-navin) [reply packages/runtime/container-runtime/src/summaryFormat.ts:128]: Also the type says `dimensions` and the property says `info`. Can we make them the same?

File: packages/runtime/container-runtime/src/garbageCollectionHelpers.ts [hunk 5]
```diff
@@ -35,3 +38,19 @@ export function sendGCUnexpectedUsageEvent(
...
+/**
+ * In order to protect old documents that were created at a time when known bugs exist that violate GC's invariants
+ * such that enforcing GC (Fail on Tombstone load/usage, GC Sweep) would cause legitimate data loss,
+ * the container author may increment the feature support value for Tombstone such that containers created
+ * with a different value will not be subjected to GC enforcement.
+ * @param persistedValue - The persisted feature support value
+ * @param currentValue - The current app-provided feature support value
+ * @returns true if GC Enforcement (Fail on Tombstone load/usage, GC Sweep) should be disabled
+ */
+export function shouldDisableGcEnforcement(
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/garbageCollectionHelpers.ts:51]: I am wondering if we should use enable instead of disable to indicate if a feature is enabled or not. I guess it depends on what we expect the primary use of GcFeatureSupportDimensions is - a means to enable features or disable them.

File: packages/runtime/container-runtime/src/summaryFormat.ts (hunk 4 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/summaryFormat.ts:128]: Ahh "gcFeature" sounds like "the GC feature", not "related features".  OK

File: packages/runtime/container-runtime/src/garbageCollectionHelpers.ts (hunk 5 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollectionHelpers.ts:51]: Yeah I got the same feedback from Tyler.  I'll do it in a quick follow-up PR

File: packages/runtime/container-runtime/src/summaryFormat.ts [hunk 6]
```diff
@@ -97,15 +97,35 @@ export interface ICreateContainerMetadata {
 	createContainerTimestamp?: number;
 }
 
+/** @see IGCMetadata.gcFeatureSupportInfo */
+export interface GcFeatureSupportDimensions {
+	/**
+	 * Gives a way for an app to disqualify old files from GC Tombstone enforcement
+	 * Provided via Container Runtime Options
+	 */
+	appTombstoneReadiness?: number;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/summaryFormat.ts:106]: I would suggest just using "tombstone" to indicate this is for the tombstone feature. Are you expecting there to be different properties for application vs FF?

File: packages/runtime/container-runtime/src/test/gcHelpers.spec.ts [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	describe("shouldDisableGcEnforcement", () => {
+		const testCases: {
+			persisted: number | undefined;
+			current: number | undefined;
+			expectedShouldDisableValue: boolean;
+		}[] = [
+			{
+				persisted: undefined,
+				current: undefined,
+				expectedShouldDisableValue: false,
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/test/gcHelpers.spec.ts:19]: I would expect this to be true.

File: packages/runtime/container-runtime/src/test/gcHelpers.spec.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+			},
+			{
+				persisted: 1,
+				current: 1,
+				expectedShouldDisableValue: false,
+			},
+			{
+				persisted: 1,
+				current: 2,
+				expectedShouldDisableValue: false,
```
[the remark this answers is not in the document]
Comment (markfields) [reply packages/runtime/container-runtime/src/test/gcHelpers.spec.ts]: Yeah the tests failed, congrats on finding the bug :)  It's fixed.

File: packages/test/test-end-to-end-tests/package.json [hunk 9]
```diff
@@ -49,7 +49,7 @@
 		"test:realsvc:tinylicious": "npm run test:realsvc:t9s",
 		"test:realsvc:tinylicious:report": "cross-env FLUID_TEST_REPORT=1 npm run test:realsvc:t9s",
 		"test:realsvc:tinylicious:run": "npm run test:realsvc:run -- --driver=t9s --timeout=5s",
-		"test:realsvc:verbose": "cross-env FLUID_TEST_VERBOSE=1 npm run test:realsvc"
+		"test:realsvc:verbose": "cross-env FLUID_TEST_VERBOSE=1 npm run test:realsvc:local"
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/package.json:52]: why this change?

File: packages/runtime/container-runtime/src/test/gcHelpers.spec.ts (hunk 7 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/test/gcHelpers.spec.ts:19]: Yeah I wrestled with this.  Keeping it simple for now, it will not hinder the immediate use case our partners have.

I think we may want persisted `undefined` value to be mapped to 0 first, then compare with the incoming current value from the option. See https://dev.azure.com/fluidframework/internal/_workitems/edit/3424 for more details.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts [hunk 10]
```diff
@@ -602,6 +625,52 @@ describeNoCompat("GC data store tombstone tests", (getTestObjectProvider) => {
...
+					await summarizationWithUnreferencedDataStoreAfterTime(sweepTimeoutMs);
+				await sendOpToUpdateSummaryTimestampToNow(summarizingContainer);
+
+				// The datastore should be tombstoned now
+				const { summaryVersion } = await summarize(summarizer);
+				const container = await loadContainer(
+					summaryVersion,
+					true /* disableTombstoneFailureViaOption */,
+				);
+				await sendOpToUpdateSummaryTimestampToNow(container);
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:655]: This should not be required.

File: packages/runtime/container-runtime/src/garbageCollectionHelpers.ts (hunk 2 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollectionHelpers.ts:55]: Closing this thread, let's use the other one from the test file.

File: packages/runtime/container-runtime/src/test/gcHelpers.spec.ts (hunk 7 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/test/gcHelpers.spec.ts:19]: In the other thread you said:

> why not return false if persistedValue is undefined? Before applications start passing the GC option, this would be undefined and tombstone will be enabled right?

I don't want this change to disable tombstone until someone wants it disabled.

File: packages/test/test-end-to-end-tests/package.json (hunk 9 above)
Comment (markfields) [reply packages/test/test-end-to-end-tests/package.json:52]: Oops

File: packages/runtime/container-runtime/src/summaryFormat.ts (hunk 4 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/summaryFormat.ts:128]: Going with gcFeatureMatrix

File: packages/runtime/container-runtime/src/summaryFormat.ts (hunk 6 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/summaryFormat.ts]: Went with `tombstoneGeneration` after discussion

File: packages/runtime/container-runtime/src/garbageCollection.ts [hunk 11]
```diff
@@ -434,6 +436,9 @@ export class GarbageCollector implements IGarbageCollector {
 	// This is the version of GC data in the latest summary being tracked.
 	private latestSummaryGCVersion: GCVersion;
 
+	// Feature Support info persisted to this container's summary
+	private readonly persistedGcFeatureSupportInfo: GCFeatureMatrix | undefined;
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/garbageCollection.ts:440]:
```suggestion
	private readonly persistedGcFeatureMatrix: GCFeatureMatrix | undefined;
```

Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollection.ts:440]: Will do in subsequent PR that flips disable -> allow nomenclature

File: packages/runtime/container-runtime/src/garbageCollectionConstants.ts [hunk 12]
```diff
@@ -10,6 +10,9 @@ export const stableGCVersion: GCVersion = 1;
 /** The current version of garbage collection. */
 export const currentGCVersion: GCVersion = 2;
 
+/** This undocumented GC Option (on ContainerRuntime Options) allows an app to disable enforcing GC on old documents by incrementing this value */
+export const gcTombstoneEnforcementValueOptionName = "gcTombstoneEnforcementValue";
```
Comment (agarwal-navin) [inline packages/runtime/container-runtime/src/garbageCollectionConstants.ts:14]: gcTombstoneGeneration?

Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollectionConstants.ts:14]: Oooh yeah gotta change this one.  Thanks.

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 13]
```diff
@@ -225,6 +225,7 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
 		// Read the feature flag that tells whether to throw when a tombstone blob is requested.
 		this.throwOnTombstoneLoad =
 			this.mc.config.getBoolean(throwOnTombstoneLoadKey) === true &&
+			!(this.runtime as ContainerRuntime).disableGcTombstoneEnforcement &&
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/blobManager.ts:228]: Is modifying IBlobManagerRuntime to include the disableGCTombstoneEnforcement flag possible/relatively easy instead of importing ContainerRuntime? Somehow, I have a feeling there's going to be a circular dependency here.

Comment (markfields) [reply packages/runtime/container-runtime/src/blobManager.ts:228]: I'll check

File: packages/runtime/container-runtime/src/blobManager.ts [hunk 14]
```diff
@@ -786,6 +787,8 @@ export class BlobManager extends TypedEventEmitter<IBlobManagerEvents> {
 						: "GC_Deleted_Blob_Requested",
 				category: shouldFail ? "error" : "generic",
 				isSummarizerClient: this.runtime.clientDetails.type === summarizerClientType,
+				gcEnforcementDisabled: (this.runtime as ContainerRuntime)
+					.disableGcTombstoneEnforcement,
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/blobManager.ts:791]: Why are we logging just the runtime level event instead of throwOnTombstoneLoad?

File: packages/runtime/container-runtime/src/garbageCollection.ts [hunk 15]
```diff
@@ -434,6 +436,9 @@ export class GarbageCollector implements IGarbageCollector {
 	// This is the version of GC data in the latest summary being tracked.
 	private latestSummaryGCVersion: GCVersion;
 
+	// Feature Support info persisted to this container's summary
+	private readonly persistedGcFeatureMatrix: GCFeatureMatrix | undefined;
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/garbageCollection.ts:440]: nit: persistedGCFeatureMatrix

File: packages/runtime/container-runtime/src/containerRuntime.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1040]: Yeah I'm pondering this too.  Need to "get it right" because I can't rename it later (breaking change).  "Enable" sounds too authoritative, since other checks have to pass too. But "allow" may be clearer.

File: packages/runtime/container-runtime/src/garbageCollection.ts (hunk 15 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollection.ts:440]: Since Navin also agrees that `Gc` is better (I like it better) I'm leaving it.  We can take a pass to clean up if we care... I don't....

File: packages/runtime/container-runtime/src/garbageCollection.ts [hunk 16]
```diff
@@ -607,6 +613,11 @@ export class GarbageCollector implements IGarbageCollector {
 			}
 			this.sweepTimeoutMs =
 				testOverrideSweepTimeoutMs ?? computeSweepTimeout(this.sessionExpiryTimeoutMs);
+			if (this.gcOptions[gcTombstoneGenerationOptionName] !== undefined) {
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/garbageCollection.ts:616]: I wonder if we make this.persistedGCFeatureMatrix always defined, that way this if statement can be simplified to a ??.

Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollection.ts:616]: It breaks snapshot tests.  I have a follow-up item opened.

File: packages/runtime/container-runtime/src/garbageCollection.ts [hunk 17]
```diff
@@ -1408,6 +1420,7 @@ export class GarbageCollector implements IGarbageCollector {
 					isSummarizerClient: this.isSummarizerClient,
 					url: trimLeadingSlashes(toNodePath),
 					nodeType,
+					gcEnforcementDisabled: false,
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/garbageCollection.ts:1423]: False gcEnforcementDisabled means it's enabled? Why is it always enabled here? And shouldn't this be true in a sense since tombstoneEnforcement is not true?

File: packages/runtime/container-runtime/src/test/gcHelpers.spec.ts (hunk 7 above)
Comment (tyler-cai-microsoft) [reply packages/runtime/container-runtime/src/test/gcHelpers.spec.ts:19]: Just doing some mental translating: Tombstone should be enabled until someone wants it disabled, right?

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts (hunk 10 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:655]: Curious as to how this would look

File: packages/runtime/container-runtime/src/garbageCollection.ts (hunk 17 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollection.ts:1423]: I think this is leftover from when I was also covering Sweep.  Good catch, I'll remove.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts (hunk 10 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneDataStores.spec.ts:655]: There's no need to update the summary timestamp to now as we are not summarizing :)

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 18]
```diff
@@ -1085,8 +1092,22 @@ export class ContainerRuntime
...
+			this.runtimeOptions.gcOptions[gcTombstoneGenerationOptionName],
+		);
+
 		this.mc = loggerToMonitoringContext(ChildLogger.create(this.logger, "ContainerRuntime"));
 
+		this.mc.logger.sendTelemetryEvent({
+			eventName: "GCFeatureMatrix",
+			info: JSON.stringify(metadata?.gcFeatureMatrix),
+			inputs: JSON.stringify({
+				gcOptions_gcTombstoneGeneration:
```
Comment (tyler-cai-microsoft) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1106]: Why is there an _ here?

Comment (markfields) [reply packages/runtime/container-runtime/src/containerRuntime.ts:1106]: Readability?  Because I can't do `.`

File: packages/runtime/container-runtime/src/garbageCollection.ts (hunk 17 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollection.ts:1423]: Umm wrong, I need to pass in the flag from ContainerRuntime into here

File: packages/runtime/container-runtime/src/test/gcHelpers.spec.ts (hunk 7 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/test/gcHelpers.spec.ts:19]: Yes.  I think I'm confident enough now to make the change to say if input is undefined, always allow (don't disable)

File: packages/runtime/container-runtime/src/blobManager.ts (hunk 13 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/blobManager.ts:228]: No circular dependency since it's all in the same pacakge.

File: packages/runtime/container-runtime/src/blobManager.ts (hunk 14 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/blobManager.ts]: throwOnTombstoneLoad is generally encapsulated in the category

File: packages/runtime/container-runtime/src/garbageCollectionHelpers.ts (hunk 5 above)
Comment (markfields) [reply packages/runtime/container-runtime/src/garbageCollectionHelpers.ts]: Did it now.

**Review by agarwal-navin — APPROVED** [review]
